py/objint.c: Add `signed` param to int.from_bytes().

IhorNehrutsa opened the pull request:
### Summary
Support `signed=True` param:
result = int.from_bytes(bytearray(), byteorder='big'|'little', signed=False|True)
```
>>> int.from_bytes(b'\xff', byteorder='little', signed=False)
255
>>> int.from_bytes(b'\xff', byteorder='little', signed=True)
-1
>>> 
```

### Testing
[test_int.py](https://github.com/micropython/micropython/pull/16798#issuecomment-2676437400)
Added:
micropython/tests/basics/int_from_bytes.py

Tested on ESP32, ESP32-S2 boards.

### Trade-offs and Alternatives
The length of the byte array is limited to `sizeof(mp_longint_impl_t)` in this PR.
int.from_bytes(bytes(20)) will raise OverflowError: big-int overflow


Wait for #16311

IhorNehrutsa commented:
```
# test_int.py


def test(x, order, signed):
    size = 4 if x < 2**30 and x >= -(2**30) else 8
    b = x.to_bytes(size, order, signed)
    y = int.from_bytes(b, order, signed)
    print(x, y, x == y, b)
    print()
    assert x == y


def do_test(n, order, signed):
    print("==========", "n:", n, "order:", order, "signed:", signed, "==========")
    test(0, order, signed)
    test(1, order, signed)
    test(2, order, signed)
    test(0xFF, order, signed)
    test(0xFFFF, order, signed)
    test(0xFFFFFF, order, signed)
    test(0x3FFFFFFF, order, signed)
    # test(0x7FFFffff, order, signed)
    # test(0xFFFFffff, order, signed)
    # test(0xffFFFFffff, order, signed)
    # test(0xffffFFFFffff, order, signed)
    # test(0xFFffffFFFFffff, order, signed)
    # test(0xFFFFffffFFFFffff, order, signed)

    if signed:
        print("----------", "n:", n, "order:", order, "signed:", signed, "----------")
        test(-0, order, signed)
        test(-1, order, signed)
        test(-2, order, signed)
        test(-0xFF - 1, order, signed)
        test(-0xFFFF - 1, order, signed)
        test(-0xFFFFFF - 1, order, signed)
        test(-0x3FFFFFFF - 1, order, signed)
        # test(-0x7FFFffff-1, order, signed)
        # test(-0xFFFFffff-1, order, signed)
        # test(-0xffFFFFffff-1, order, signed)
        # test(-0xffffFFFFffff-1, order, signed)
        # test(-0xFFffffFFFFffff-1, order, signed)
        # test(-0xFFFFffffFFFFffff-1, order, signed)


do_test(n=1, order="little", signed=False)
do_test(n=2, order="big", signed=False)
do_test(n=3, order="little", signed=True)
do_test(n=4, order="big", signed=True)
print("Ok.")
```

IhorNehrutsa commented:
```
MicroPython v1.25.0-preview.308.g1f1bcb5ca.dirty on 2025-02-23; Generic ESP32S2 module with ESP32S2
>>> %Run -c $EDITOR_CONTENT
MPY: soft reboot
========== n: 1 order: little signed: False ==========
big_endian:0 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:0
0 0 True b'\x00\x00\x00\x00'

big_endian:0 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:1
1 1 True b'\x01\x00\x00\x00'

big_endian:0 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:2
2 2 True b'\x02\x00\x00\x00'

big_endian:0 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:255
255 255 True b'\xff\x00\x00\x00'

big_endian:0 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:65535
65535 65535 True b'\xff\xff\x00\x00'

big_endian:0 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:16777215
16777215 16777215 True b'\xff\xff\xff\x00'

big_endian:0 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:1073741823
1073741823 1073741823 True b'\xff\xff\xff?'

========== n: 2 order: big signed: False ==========
big_endian:1 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:0
0 0 True b'\x00\x00\x00\x00'

big_endian:1 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:1
1 1 True b'\x00\x00\x00\x01'

big_endian:1 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:2
2 2 True b'\x00\x00\x00\x02'

big_endian:1 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:255
255 255 True b'\x00\x00\x00\xff'

big_endian:1 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:65535
65535 65535 True b'\x00\x00\xff\xff'

big_endian:1 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:16777215
16777215 16777215 True b'\x00\xff\xff\xff'

big_endian:1 signed:0 bufinfo.len:4 sizeof(result.uval):4 result.uval:1073741823
1073741823 1073741823 True b'?\xff\xff\xff'

========== n: 3 order: little signed: True ==========
big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:0
0 0 True b'\x00\x00\x00\x00'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:1
1 1 True b'\x01\x00\x00\x00'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:2
2 2 True b'\x02\x00\x00\x00'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:255
255 255 True b'\xff\x00\x00\x00'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:65535
65535 65535 True b'\xff\xff\x00\x00'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:16777215
16777215 16777215 True b'\xff\xff\xff\x00'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:1073741823
1073741823 1073741823 True b'\xff\xff\xff?'

---------- n: 3 order: little signed: True ----------
big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:0
0 0 True b'\x00\x00\x00\x00'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-1
-1 -1 True b'\xff\xff\xff\xff'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-2
-2 -2 True b'\xfe\xff\xff\xff'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-256
-256 -256 True b'\x00\xff\xff\xff'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-65536
-65536 -65536 True b'\x00\x00\xff\xff'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-16777216
-16777216 -16777216 True b'\x00\x00\x00\xff'

big_endian:0 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-1073741824
-1073741824 -1073741824 True b'\x00\x00\x00\xc0'

========== n: 4 order: big signed: True ==========
big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:0
0 0 True b'\x00\x00\x00\x00'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:1
1 1 True b'\x00\x00\x00\x01'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:2
2 2 True b'\x00\x00\x00\x02'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:255
255 255 True b'\x00\x00\x00\xff'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:65535
65535 65535 True b'\x00\x00\xff\xff'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:16777215
16777215 16777215 True b'\x00\xff\xff\xff'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:1073741823
1073741823 1073741823 True b'?\xff\xff\xff'

---------- n: 4 order: big signed: True ----------
big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:0
0 0 True b'\x00\x00\x00\x00'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-1
-1 -1 True b'\xff\xff\xff\xff'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-2
-2 -2 True b'\xff\xff\xff\xfe'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-256
-256 -256 True b'\xff\xff\xff\x00'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-65536
-65536 -65536 True b'\xff\xff\x00\x00'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-16777216
-16777216 -16777216 True b'\xff\x00\x00\x00'

big_endian:1 signed:1 bufinfo.len:4 sizeof(result.ival):4 result.ival:-1073741824
-1073741824 -1073741824 True b'\xc0\x00\x00\x00'

Ok.
>>> 
```

IhorNehrutsa commented:
@projectgus @dpgeorge @robert-hh 

Need a help.
Is it possible to use MICROPY_LONGINT_IMPL_LONGLONG?
```
/home/runner/work/micropython/micropython/ports/esp32/build-ESP32_GENERIC/frozen_content.c:22:2: error: #error "incompatible MICROPY_LONGINT_IMPL"
   22 | #error "incompatible MICROPY_LONGINT_IMPL"
      |  ^~~~~
```
Thanks.

projectgus commented:
> Support `signed=True` param: result = int.from_bytes(bytearray(), order='big'|'little', signed=False|True)

See also https://github.com/micropython/micropython/pull/16311

projectgus commented:
> Is it possible to use MICROPY_LONGINT_IMPL_LONGLONG?

I think the error you have here may be caused by a partial build failure, you can try doing a clean build and see if it goes away. However I'm not certain every longint implementation works on every port.

IhorNehrutsa commented:
> caused by a partial build failure, you can try doing a clean build

```
cd ~/micropython/ports/esp32
make clean

rm -R managed_components
rm -R build*
```

The same errors remain. :(

projectgus commented:
> The same errors remain. :(

As well as changing the constant in `mpconfigport.h`, you'll need to change the build system so `mpy-tool.py` is run with the extra argument `-mlongint_impl=longlong`. 

I think for esp32 port this will mean something like adding `set(MICROPY_MPY_TOOL_FLAGS -mlongint_impl=longlong)` in the CMake file. I haven't tested this, it's not really a supported configuration.

IhorNehrutsa committed:
py/objint: Add signed param to int.from_bytes().

Support signed param:
result = int.from_bytes(bytearray(),
byteorder='big'|'little', signed=False|True)

Signed-off-by: Ihor Nehrutsa <Ihor.Nehrutsa@gmail.com>

IhorNehrutsa commented:
Ready for code review

dpgeorge commented:
@IhorNehrutsa there is a lot of failing CI that needs to be addressed.

IhorNehrutsa commented:
Cherry pick commit to #16311

IhorNehrutsa commented:
I don't understand where is the problem. Need a help.
```
Run tools/ci.sh unix_coverage_run_tests
make: Entering directory '/home/runner/work/micropython/micropython/ports/unix'
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
Including User C Module from ../../examples/usercmodule/cexample
Including User C Module from ../../examples/usercmodule/cppexample
Including User C Module from ../../examples/usercmodule/subpackage
cd ../../tests && MICROPY_MICROPYTHON=/home/runner/work/micropython/micropython/ports/unix/build-coverage/micropython ./run-tests.py
platform=linux arch=x64 thread=unsafe float=64-bit unicode
pass  basics/0prelim.py 
...
pass  unicode/unicode_ure.py 
1097 tests performed (33170 individual testcases)
1096 tests passed
20 tests skipped:
1 tests failed: micropython/heapalloc_int_from_bytes.py
```

~/micropython/ports/unix$ ./build-standard/micropython
```
MicroPython v1.29.0-preview.424.g2d99bb343b on 2026-06-26; linux [GCC 13.3.0] version
Type "help()" for more information.
>>> import micropython
>>> micropython.heap_lock()
>>> int.from_bytes(b"1", "little")
49
>>> int.from_bytes(b"12", "little")
12849
>>> int.from_bytes(b"2\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00", "little")
50
>>> micropython.heap_unlock()
32767
>>>
```
The data output generally matches the sample file:
micropython\tests\micropython\heapalloc_int_from_bytes.py.exp 
```
49
12849
50
```
???

dpgeorge commented:
When you use the REPL, `micropython.heap_unlock()` is automatically called at each line, so testing the heap-lock code at the REPL will not work (unless you use paste mode).

It looks like `int.from_bytes()` is allocating when it should not be.

dpgeorge commented on py/objint.c:
```diff
@@ -398,37 +398,31 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
-        #endif
-        value = (value << 8) | *buf;
-    }
-    return mp_obj_new_int_from_uint(value);
```
This code path needs to stay, so it can create a small int if the result will fit in a small int.

IhorNehrutsa committed:
py/objint: Revert to using small_int.

Signed-off-by: Ihor Nehrutsa <Ihor.Nehrutsa@gmail.com>

IhorNehrutsa commented:
```
$ make clean VARIANT=minimal
$ make VARIANT=minimal
```
```
LINK build-minimal/micropython
   text    data     bss     dec     hex filename
 178237   14712    1248  194197   2f695 build-minimal/micropython
```
```
$ ./build-minimal/micropython  ../../tests/basics/int_from_bytes.py
```
```
Traceback (most recent call last):
  File "../../tests/basics/int_from_bytes.py"
SyntaxError: invalid syntax
```

IhorNehrutsa commented:
```
./build-longlong/micropython  ../../tests/basics/int_from_bytes.py
```
```
...
x:-4611686018427387902, x:0x-3FFFFFFFFFFFFFFE, size:8, to_bytes:b'\xc0\x00\x00\x00\x00\x00\x00\x02', y = x.from_bytes:-4611686018427387902, True

x:-4611686018427387904, x:0x-4000000000000000, size:8, to_bytes:b'\xc0\x00\x00\x00\x00\x00\x00\x00', y = x.from_bytes:-4611686018427387904, True

x:-4611686018427387906, x:0x-4000000000000002, size:12, to_bytes:b'\xff\xff\xff\xff\xbf\xff\xff\xff\xff\xff\xff\xfe', y = x.from_bytes:-4611686018427387906, True

Ok.
```

IhorNehrutsa commented:
@projectgus 
Can you review the code?

IhorNehrutsa committed:
Update objint.c

dpgeorge commented on tests/basics/int_from_bytes.py:
```diff
@@ -0,0 +1,151 @@
+    elif x < 2**62 and x >= -(2**62):
+        size = 8
+    
+    print(f"x:{x}, x:0x{x:X}, size:{size}", end="")
```
Please don't use f-strings in tests, use .format() instead (that's partly why the CI is failing).

projectgus commented on docs/library/builtins.rst:
```diff
@@ -74,9 +74,9 @@ Functions and types
 
-      In MicroPython, `byteorder` parameter must be positional (this is
+      In MicroPython, `byteorder` parameter may be positional (this is
       compatible with CPython).
```
IIUC this PR will make this the same as CPython, so this note can be removed?

projectgus commented on py/mpz.c:
```diff
@@ -850,37 +850,56 @@ size_t mpz_set_from_str(mpz_t *z, const char *str, size_t len, bool neg, unsigne
+        z->dig[dig_idx++] = d & DIG_MASK;
+    }
+    z->len = mpn_remove_trailing_zeros(z->dig, z->dig + dig_idx);
+    if (z->len == 1 && z->dig[0] == 0) {
+        z->len = 0;
+        z->neg = 0;
+    }
```
Do we need this? The buffer is allocated regardless, so we might be able to save a little bit of code size by not doing this check?

projectgus commented on py/objint.c:
```diff
@@ -398,37 +398,48 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
+    bool is_signed = args[ARG_signed].u_bool;
 
-    const byte *buf = (const byte *)bufinfo.buf;
+    const byte *buf = (const byte *)bufinfo.buf; // buf is the MSB for big-endian
```
```suggestion
    const byte *buf = (const byte *)bufinfo.buf;
```

(same comment is added on line 419, where the change is applied.)

projectgus commented on py/objint_longlong.c:
```diff
@@ -39,6 +39,9 @@
 
+#define MP_LONG_LONG_MAX LLONG_MAX
+#define MP_LONG_LONG_MIN LLONG_MIN
+
```
`LLONG_MAX` and `LLONG_MIN` are in C99 so can probably just use them as-is, should be supported on all compilers

projectgus commented on tests/basics/int_from_bytes.py:
```diff
@@ -0,0 +1,151 @@
+    test_bytes(b"\x00\x00\x00\x00\x00\x00\x02", 2 << 48, byteorder, signed)
+    test_bytes(b"\x00\x00\x00\x00\x00\x00\xFE", 254 << 48, byteorder, signed)
+    # test_bytes(b"\x00\x00\x00\x00\x00\x00\x00\x02", 2 << 56, byteorder, signed)
+    # test_bytes(b"\x00\x00\x00\x00\x00\x00\x00\xFE", 254 << 56, byteorder, signed)
```
Why are these two commented?

projectgus commented on py/objint.c:
```diff
@@ -398,37 +398,48 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_binary_op_t op, mp_obj_t lhs_in, mp
+                return mp_obj_int_from_bytes_impl(big_endian, is_signed, bufinfo.len, bufinfo.buf);
+            }
+            #endif
+            value = (value << 8) | *buf;
```
Left shift of a negative value (created at line 426) is undefined behaviour in C. 

This was also caught by the CI `sanitize_undefined` job: `../../py/objint.c:437:28: runtime error: left shift of negative value -2`.

https://github.com/micropython/micropython/actions/runs/28425122420/job/84226398605

projectgus reviewed (commented):
Hi @IhorNehrutsa,

Thanks for persisting with this PR. It looks like a potentially useful addition.

I've left some comments on the code but the more important issue with this PR is the failing CI. Several of these failures are pointing at real problems that need to be resolved. Look at the "Print failures" section of each job log for details.

The coverage report also shows some lines aren't covered by tests yet (although the coverage tool page isn't loading for me, so I can't see which lines it is.)

If you get stuck on debugging a particular CI failure then feel free to ask for help with it. We can't do a full code review until the behaviour is correct as verified by CI.

As well as the failing CI, I have some concern about the code size increase. I'm not sure if there's anything that can be done to reduce this.

IhorNehrutsa committed:
int_from_bytes.py: Use format().

Signed-off-by: Ihor Nehrutsa <Ihor.Nehrutsa@gmail.com>

IhorNehrutsa committed:
Update builtins.rst

IhorNehrutsa committed:
Update mpz.c

IhorNehrutsa committed:
Update objint.c

IhorNehrutsa committed:
Update objint_longlong.c

IhorNehrutsa committed:
Update objint.h

IhorNehrutsa committed:
Update mpz.c

IhorNehrutsa committed:
Update int_from_bytes.py

projectgus commented:
@IhorNehrutsa I'm going to refer to my previous comments again:

> I've left some comments on the code but the more important issue with this PR is the failing CI.
> [...]
> The coverage report also shows some lines aren't covered by tests yet (although the coverage tool page isn't loading for me, so I can't see which lines it is.)
>
> If you get stuck on debugging a particular CI failure then feel free to ask for help with it. We can't do a full code review until the behaviour is correct as verified by CI.

If you've got specific questions about the code, then please ask. However, please respect my time by fixing CI before asking for another full review.

IhorNehrutsa commented:
I can't resolve CI. May you give some hints?

IhorNehrutsa committed:
Update int_from_bytes.py

dpgeorge commented:
@IhorNehrutsa you need to try and fix the CI yourself, by clicking through and looking at the errors.  They give more feedback than we can give at this stage.  Eg:
```
-bytes:b'\x00\x00\x00\x00\xfe', size:5, byteorder:little, signed:True, sint:-8589934592, from_bytes:-8589934592, True, to_bytes:b'\x00\x00\x00\x00\xfe'
+bytes:b'\x00\x00\x00\x00\xfe', size:5, byteorder:little, signed:True, ../../py/objint_longlong.c:68:28: runtime error: left shift of negative value -2
+sint:-8589934592, from_bytes:-8589934592, True, to_bytes:b'\x00\x00\x00\x00\xfe'
 ```
And:
```
+bytes:b'\x00', size:1, byteorder:big, signed:False, Traceback (most recent call last):
+  File "/home/runner/work/micropython/micropython/tests/basics/int_from_bytes.py", in <module>
+  File "/home/runner/work/micropython/micropython/tests/basics/int_from_bytes.py", in do_test_bytes
+  File "/home/runner/work/micropython/micropython/tests/basics/int_from_bytes.py", in test_bytes
+  File "/home/runner/work/micropython/micropython/tests/basics/int_from_bytes.py", in reverse
+NameError: name not defined
```
For the latter: the unix minimal port doesn't have `bytearray`.  So to implement the `reverse()` function I suggest using bytes with list:
```py
def reverse(b):
    r = []
    for x in b:
        r.insert(0, x)
    return bytes(r)
```

dpgeorge commented on py/mpz.c:
```diff
@@ -850,37 +850,60 @@ size_t mpz_set_from_str(mpz_t *z, const char *str, size_t len, bool neg, unsigne
-
-    z->len = mpn_remove_trailing_zeros(z->dig, z->dig + z->len);
+    if (num_bits > 0 && dig_idx < z->alloc) {
+        z->dig[dig_idx++] = d & DIG_MASK;
```
Codecov is showing that this line is not tested.  It needs a test.

IhorNehrutsa committed:
Update int_from_bytes.py

IhorNehrutsa committed:
Update objint_longlong.c

dpgeorge commented:
The tests on longlong and minimal builds are now raising OverflowError.  That's probably because the test is trying to convert bytes that are too large to be represented as an int on those builds (which do not have arbitrary sized integers).

You'll need to split the new test file into three:
- `int_from_bytes.py` that only tests things resulting in small-ints (less than 2**30)
- `int_from_bytes_int64.py` that tests up to long-long size (2**63)
- `int_from_bytes_intbig.py` that has the larger tests

IhorNehrutsa committed:
Update int_from_bytes.py

IhorNehrutsa committed:
Create int_from_bytes_int64.py

IhorNehrutsa committed:
Create int_from_bytes_intbig.py

IhorNehrutsa commented:
Must I to squash commits?

dpgeorge commented:
> Must I to squash commits?

Eventually, yes.  But it's not needed yet.

I made a comment above about a line in mpz.c not being covered by a test.  You need to respond to that comment.  There are also 4 other lines that codecov is showing as uncovered.  You need to write tests that execute those lines.  If you can't write such a test then that means those lines of code are incorrect and need to be removed.

dpgeorge commented on py/mpz.c:
```diff
@@ -850,37 +850,60 @@ size_t mpz_set_from_str(mpz_t *z, const char *str, size_t len, bool neg, unsigne
+        return;
+    }
+    if (is_signed && (buf[big_endian ? 0 : len - 1] & 0x80)) {
+        z->neg = 1;
```
Codecov says this line is uncovered by tests.

dpgeorge commented on py/mpz.c:
```diff
@@ -850,37 +850,60 @@ size_t mpz_set_from_str(mpz_t *z, const char *str, size_t len, bool neg, unsigne
+        if (z->neg) {
+            byte_val = ((~byte_val) & 0xff) + carry;
+            carry = byte_val >> 8;
+            byte_val &= 0xff;
```
Codecov says these 3 lines above are uncovered by tests.

IhorNehrutsa commented:
Before splitting int_from_bytes.py to three files only 897-898 in py/mpz.c was in codecow report.
After splitting into three files, other lines were appeared.
I don't know how to solve this.

dpgeorge commented on py/mpz.c:
```diff
@@ -850,37 +850,60 @@ size_t mpz_set_from_str(mpz_t *z, const char *str, size_t len, bool neg, unsigne
-            num_bits += 8;
-            buf += delta;
-            len--;
+        while (num_bits >= DIG_SIZE) {
```
I don't understand why this while loop is here: `num_bits` can only grow by 8, which is never larger than `DIG_SIZE`.  So this loop will execute at most once.

Please try to minimise the diff of this function compared to what it is currently in master.  That will make it easier to review and reason about.

IhorNehrutsa committed:
Update mpz.c

IhorNehrutsa committed:
Update int_from_bytes.py

IhorNehrutsa committed:
Update int_from_bytes.py

IhorNehrutsa committed:
Update int_from_bytes_int64.py

IhorNehrutsa committed:
Delete int_from_bytes_int64.py

IhorNehrutsa committed:
Delete int_from_bytes_intbig.py

IhorNehrutsa commented:
**longlong:** Where is the error line?
https://github.com/micropython/micropython/actions/runs/28927391058/job/85818073291?pr=16798
'''
Run tools/ci.sh unix_longlong_run_tests
make: Entering directory '/home/runner/work/micropython/micropython/ports/unix'
Use make V=[1](https://github.com/micropython/micropython/actions/runs/28927391058/job/85818073291?pr=16798#step:6:1) or set BUILD_VERBOSE in your environment to increase build verbosity.
cd ../../tests && MICROPY_MICROPYTHON=/home/runner/work/micropython/micropython/ports/unix/build-longlong/micropython ./run-tests.py
platform=linux arch=x86 thread=unsafe float=30-bit unicode
pass  basics/0prelim.py 
pass  basics/equal_class.py
...
pass  unicode/unicode_subscr.py 
skip  unicode/unicode_ure.py
971 tests performed (21700 individual testcases)
970 tests passed
152 tests skipped: basics/builtin_pow3_intbig.py basics/builtin_range_binop.py basics/floordivide_intbig.py basics/class_inplace_op2.py basics/builtin_round_intbig.py basics/array_intbig.py basics/array_limits_intbig.py basics/attrtuple2.py basics/bytearray_intbig.py basics/builtin_abs_intbig.py basics/builtin_bin_intbig.py basics/fun_code.py basics/fun_code_colines.py basics/fun_code_full.py basics/fun_code_lnotab.py basics/builtin_divmod_intbig.py basics/bytes_construct_intbig.py basics/builtin_hash_intbig.py basics/builtin_hex_intbig.py basics/builtin_oct_intbig.py basics/struct1_intbig.py basics/slice_intbig.py basics/int_big1.py basics/int_big_add.py basics/int_big_and.py basics/int_big_and2.py basics/int_big_and3.py basics/int_big_cmp.py basics/int_big_div.py basics/int_big_error.py basics/int_big_lshift.py basics/int_big_mod.py basics/int_big_mul.py basics/int_big_or.py basics/int_big_or2.py basics/int_big_or3.py basics/int_big_pow.py basics/int_big_rshift.py basics/int_big_to_small.py basics/int_big_to_small_int29.py basics/int_big_unary.py basics/int_big_xor.py basics/int_big_xor2.py basics/int_big_xor3.py basics/int_big_zeroone.py basics/special_methods_intbig.py basics/int_bytes_intbig.py basics/int_constfolding_intbig.py basics/int_divmod_intbig.py basics/int_intbig.py basics/io_buffered_writer.py basics/sys_getsizeof.py basics/memoryview_intbig.py basics/sys_path.py basics/memoryview_itemsize.py basics/sys_tracebacklimit.py basics/string_format_intbig.py basics/namedtuple_asdict.py basics/weakref_ref_collect.py basics/nanbox_smallint.py basics/op_error_intbig.py basics/string_module_tstring.py cmdline/cmd_parsetree.py basics/string_tstring_basic.py basics/string_tstring_basic1.py basics/string_tstring_constructor.py basics/string_tstring_constructor1.py basics/string_tstring_errors1.py basics/string_tstring_format1.py basics/string_tstring_interpolation1.py basics/string_tstring_operations.py basics/string_tstring_parser1.py basics/string_tstring_whitespace.py extmod/cryptolib_aes128_ctr.py basics/weakref_callback_exception.py basics/weakref_finalize_basic.py basics/weakref_finalize_collect.py basics/weakref_multiple_refs.py basics/weakref_ref_basic.py extmod/deflate_compress.py extmod/deflate_compress_memory_error.py extmod/deflate_stream_error.py extmod/re_groups.py extmod/re_span.py extmod/machine_disable_irq.py extmod/machine_i2s_rate.py extmod/machine_rtc.py extmod/machine_spi_rate.py extmod/machine_timer.py extmod/machine_uart_irq_txidle.py extmod/machine_uart_tx.py extmod/marshal_basic.py extmod/marshal_fun_nested.py extmod/marshal_stress.py extmod/re_debug.py float/math_factorial_intbig.py float/math_fun_intbig.py float/string_format_modulo2_intbig.py import/import_file.py extmod/ssl_ioctl.py extmod/ssl_poll.py io/argv.py micropython/builtin_tstring.py micropython/const_intbig.py micropython/heap_locked.py micropython/heapalloc_fail_tstring.py micropython/import_mpy_native_gc.py micropython/native_const_intbig.py micropython/native_fun_attrs_code.py micropython/native_marshal.py micropython/viper_const_intbig.py micropython/viper_misc_intbig.py misc/cexample_class.py misc/cexample_module.py misc/cexample_subclass.py misc/sys_settrace_cov.py misc/sys_settrace_features.py misc/sys_settrace_generator.py misc/sys_settrace_loop.py ports/unix/extra_coverage.py ports/unix/ffi_callback.py ports/unix/ffi_float.py ports/unix/ffi_float2.py ports/unix/ffi_int_base.py ports/unix/ffi_int_long32.py ports/unix/ffi_int_long64.py ports/unix/ffi_int_types.py thread/disable_irq.py thread/mutate_bytearray.py thread/mutate_dict.py thread/mutate_instance.py thread/mutate_list.py thread/mutate_set.py float/builtin_float_round_intbig.py float/complex1_intbig.py cmdline/repl_sys_ps1_ps2.py float/float2int_doubleprec_intbig.py float/float2int_fp30_intbig.py float/float2int_intbig.py float/float_divmod.py float/float_format_ints_doubleprec.py float/float_format_ints_power10.py float/float_parse_doubleprec.py float/float_struct_e.py float/float_struct_e_doubleprec.py float/int_big_float.py extmod/uctypes_32bit_intbig.py extmod/uctypes_array_assign_native_le_intbig.py extmod/uctypes_array_load_store.py extmod/asyncio_as_uasyncio.py thread/thread_lock4_intbig.py unicode/unicode_ure.py
1 tests failed: basics/int_from_bytes.py
make: *** [Makefile:272: test_full_no_native] Error 1
make: Leaving directory '/home/runner/work/micropython/micropython/ports/unix'
Error: Process completed with exit code 2.
```

dpgeorge commented:
The test is failing at the start with "OverflowError: result overflows long long storage".

IhorNehrutsa commented:
It must to be skipped.
'''
pc@PC:~ /micropython/ports/unix$ ./build-longlong/micropython  ../../tests/basics/int_from_bytes.py
SKIP
pc@PC:~ /micropython/ports/unix$
'''

dpgeorge commented:
@IhorNehrutsa I appreciate your attempt here to improve the code, and thanks for your efforts so far.

But I feel that this particular improvement is too difficult.  It requires precise changes to core MPZ routines, getting 100% coverage, and making it work across all build configurations.  We are currently going back-and-forth and around in circles trying to tick all the boxes, and have been making very slow progress in that direction.  It's taking up way too much of our (maintainer) time.

So I'll close this PR.  I think you'll have more fun and more success working on a different problem.
